File: interview/Cracking the Coding Interview/chapter 9/9.5/python/9.5.py
def permutations(S):
    perms(S,S,'')

def perms(S, rem, s):
    if len(s) == len(S):
        print(s)
        return

    for i in range(0,len(rem)):
        perms(S, rem[:i]+rem[i+1:], s+rem[i])

permutations('abcd')

File: interview/Cracking the Coding Interview/chapter 9/9.5/ruby/9.5.rb
def permutations(s)
	perms(s,s,'')
end

def perms(s,rem,curr)
	if s.length == curr.length
		puts curr
		return
	end
	
	for c in rem.split('')
		perms(s,(rem.split('') - [c]).join, rem+c)
	end
end

permutations('abcd')